# Analiz

In [1]:
import pandas as pd
import numpy as np

In [2]:
data = pd.read_csv("Housing.csv")
data

id             date     price  bedrooms  bathrooms  \
0      7229300521  20141013T000000  231300.0         2       1.00   
1      6414100192  20141209T000000  538000.0         3       2.25   
2      5631500400  20150225T000000  180000.0         2       1.00   
3      2487200875  20141209T000000  604000.0         4       3.00   
4      1954400510  20150218T000000  510000.0         3       2.00   
...           ...              ...       ...       ...        ...   
21608   263000018  20140521T000000  360000.0         3       2.50   
21609  6600060120  20150223T000000  400000.0         4       2.50   
21610  1523300141  20140623T000000  402101.0         2       0.75   
21611   291310100  20150116T000000  400000.0         3       2.50   
21612  1523300157  20141015T000000  325000.0         2       0.75   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
0             1180      5650     1.0           0     0  ...      7   
1             2570      7242     2.0           0     0  ...      7   
2              770     10000     1.0           0     0  ...      6   
3             1960      5000     1.0           0     0  ...      7   
4             1680      8080     1.0           0     0  ...      8   
...            ...       ...     ...         ...   ...  ...    ...   
21608         1530      1131     3.0           0     0  ...      8   
21609         2310      5813     2.0           0     0  ...      8   
21610         1020      1350     2.0           0     0  ...      7   
21611         1600      2388     2.0           0     0  ...      8   
21612         1020      1076     2.0           0     0  ...      7   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
0            1180              0      1955             0    98178  47.5112   
1            2170            400      1951          1991    98125  47.7210   
2             770              0      1933             0    98028  47.7379   
3            1050            910      1965             0    98136  47.5208   
4            1680              0      1987             0    98074  47.6168   
...           ...            ...       ...           ...      ...      ...   
21608        1530              0      2009             0    98103  47.6993   
21609        2310              0      2014             0    98146  47.5107   
21610        1020              0      2009             0    98144  47.5944   
21611        1600              0      2004             0    98027  47.5345   
21612        1020              0      2008             0    98144  47.5941   

          long  sqft_living15  sqft_lot15  
0     -122.257           1340        5650  
1     -122.319           1690        7639  
2     -122.233           2720        8062  
3     -122.393           1360        5000  
4     -122.045           1800        7503  
...        ...            ...         ...  
21608 -122.346           1530        1509  
21609 -122.362           1830        7200  
21610 -122.299           1020        2007  
21611 -122.069           1410        1287  
21612 -122.299           1020        1357  

[21613 rows x 21 columns]

Veri kümesi 21613 satır ve 21 sütun içermektedir. Her satır, bir ev satışına ilişkin bilgileri içerir. Amacımız, diğer sütunlardaki değerleri kullanarak "price" sütunundaki değeri tahmin etmektir. Bu sayede, evin özelliklerine dayanarak yeni satışların fiyatlarını tahmin edebiliriz.

Veri kümesinde yer alan sütunlar şunlardır:

In [ ]:
# Veri kümesi hakkında bilgi
data.info()

In [ ]:
data.describe()

In [ ]:
# Eksik değerleri kontrol
data.isnull().sum()

Verilere kontrol ettiğimizde boş değerimiz bulunmamaktadır. Bu yüzden eksik verileri tamamlama adımını atlıyoruz.

Veri dağılımını inceleyelim

In [ ]:
import matplotlib.pyplot as plt
# Sayısal değişkenlerin dağılımını görselleştirme
data.hist(bins=20, figsize=(15, 10))
plt.show()

Price (fiyat) sütununun diğer sütunlar ile arasında ki ilişkiye bakalım.

In [ ]:
import seaborn as sns

# Korelasyon matrisini hesaplama
heat_data = data.drop(["date"], axis=1).corr()

# 'price' sütununun diğer sütunlarla olan korelasyonunu alma
price_corr = heat_data[['price']].drop(['id', 'price'])  # Sadece 'price' sütununu al ve 'price' sütununu çıkar

# Heatmap oluşturma
plt.figure(figsize=(12, 8))  # Grafik boyutunu ayarlama
sns.heatmap(price_corr, cmap='YlGnBu', annot=True, fmt=".2f", linewidths=.5)

# Grafik başlığı ve etiketler
plt.title('Price Sütununun Diğer Özelliklerle Korelasyonu')
plt.xlabel('Özellikler')
plt.ylabel('Korelasyon')
plt.show()

sqft_living: Yaşam alanı büyüklüğü metrekare cinsinden

grade: Genel not derecesi (1 ila 13) kategorik değişken yap iyi orta kötü

sqft_above: Zemin seviyesinden yüksek yaşam alanı metrekare cinsinden

sqft_living15: En yakın 15 mülkün yaşam alanı büyüklüğü metrekare cinsinden

bathrooms: Banyo sayısı

Price (fiyat) değerinin diğer değerler ile arasında ki ilişkiye bakıyoruz. 

sqft_living, grade, sqft_above, sqft_living15, bathrooms sütunları price sütunu ile arasında doğrusallık olduğunu gözlemliyoruz.

In [ ]:
import folium

f = folium.Figure(width=500, height=500)
map = folium.Map(location=[data.lat.mean(), data.long.mean()], zoom_start=10).add_to(f)
folium.Marker(location=[data.lat.mean(), data.long.mean()], popup='Merkez Noktamız').add_to(map)

for i in range(10):
    sample_data = data.sample()
    folium.Marker(location=[sample_data.lat, sample_data.long], popup=f'Ev {i+1}').add_to(map)

map

# Ön İşleme

In [ ]:
data = data.drop(["id"], axis=1)

In [ ]:
# Tarih sütununu yıl ve ay olarak ayır
data['date'] = pd.to_datetime(data['date'], format='%Y%m%dT%H%M%S')
data['year'] = data['date'].dt.year
data['month'] = data['date'].dt.month
data.drop('date', axis=1, inplace=True)

Tarih sütununu formatladıktan sonra yıl ve ay değerlerini ayrı sütun olarak ayırıyoruz.

In [ ]:
grade_data = data["grade"].value_counts()

Grade sütunu değer dağımlarına bakıyoruz.

In [ ]:
# Verileri x ve y eksenlerine ayırın
grades = grade_data.index
counts = grade_data.values

# Bar plot oluştur
plt.figure(figsize=(10, 6))
plt.bar(grades, counts, color='blue')

# Grafiği düzenle
plt.title('Bar Plot of Grades')
plt.xlabel('Grade')
plt.ylabel('Count')
plt.xticks(grades)  # X eksenindeki değerleri doğru şekilde etiketle
plt.grid(axis='y')

# Grafiği göster
plt.show()

Grade sütunu dağılımlara baktığımızda bazı değerlerin çok yüksek bazı değerlerin az olması sebebiyle bu değerleri kategorik veriye dönüştürmek model performansını arttıracaktır.

Kötü, Orta, İyi olarak 3 'e bölüyoruz.

In [ ]:
def categorize(value):
    if value < 7:
        return "kötü"
    elif 7 <= value <= 9:
        return "orta"
    else:
        return "iyi"

categorized_data = data["grade"].apply(categorize)
data["grade"] = categorized_data

Böldüğümüz verileri modele verebilmek için one-hot uyguluyoruz.

In [ ]:
from sklearn import preprocessing
enc = preprocessing.OneHotEncoder()
enc.fit(data[["grade"]])
enc.categories_

In [ ]:
one_hot = enc.transform(data[["grade"]]).toarray()

In [ ]:
encoded_cols = list(enc.get_feature_names_out(['grade']))
data[encoded_cols] = one_hot
data.drop(["grade"], axis=1, inplace=True)

In [ ]:
data["bathrooms"] = data["bathrooms"].apply(lambda x: int(x))

Float değerinde bathrooms olamayacağı için yuvarlama işlemi uyguluyoruz

In [ ]:
# total rooms
data["total_rooms"] = data["bathrooms"] + data["bedrooms"]

"total_rooms" gibi yeni bir özellik türetmek, modelin performansını artırabilir ve veri setindeki ilişkileri daha iyi yakalayabilir. Bu tür yeni özellikler, modelin daha doğru tahminler yapmasına ve daha anlamlı sonuçlar üretmesine yardımcı olabilir. Bu yüzden oda sayılarını topladık.

In [ ]:
outliers_columns = ["price", "sqft_living", "sqft_above", "sqft_living15", "total_rooms"]
outliers_data = data[outliers_columns]

Q1 = outliers_data.quantile(0.25)
Q3 = outliers_data.quantile(0.75)
IQR = Q3 - Q1

outliers = (outliers_data < (Q1 - 1.5 * IQR)) | (outliers_data > (Q3 + 1.5 * IQR))

# Aykırı değerlerin çıkarılması
data_cleaned = data[~outliers.any(axis=1)]
data_cleaned

In [ ]:
import seaborn as sns

heat_data = data_cleaned.corr()
price_corr = heat_data[['price']].drop(['price'])  # Sadece 'price' sütununu al ve 'price' sütununu çıkar

plt.figure(figsize=(12, 8))  # Grafik boyutunu ayarlama
sns.heatmap(price_corr, cmap='YlGnBu', annot=True, fmt=".2f", linewidths=.5)

plt.title('Price Sütununun Diğer Özelliklerle Korelasyonu')
plt.xlabel('Özellikler')
plt.ylabel('Korelasyon')
plt.show()

# Model

In [ ]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, mean_absolute_percentage_error, r2_score

def error_calculator(test, pred):
  # MAE
  mae = mean_absolute_error(test, pred)
  # MSE
  mse = mean_squared_error(test, pred, squared=False)
  # RMSE
  rmse = np.sqrt(mse)
  # MAPE
  mape = mean_absolute_percentage_error(test, pred) * 100
  # R2
  r2 = r2_score(test, pred)

  return f"{mae:.2f}", f"{mse:.2f}", f"{rmse:.2f}", mape, r2

In [ ]:
def pred_plot(test, pred, model = ""):
  # Figure ve Subplot ayarları
  plt.figure(figsize=(14, 8))
  # İkinci Subplot (Scatter Plot)
  plt.subplot(2, 1, 2)
  plt.scatter(range(len(test)), test, color='blue', label='Gerçek Değerler', alpha=1)
  plt.scatter(range(len(pred)), pred, color='red', label='Tahmin Değerleri', alpha=0.3, marker="x")
  plt.title(f'({model}) Gerçek ve Tahmin Değerleri')
  plt.xlabel('Fiyat')
  plt.ylabel('Değer')
  plt.legend()
  plt.grid(True)

  # Grafiği göster
  plt.tight_layout()
  plt.show()

In [ ]:
# Özellikler (X) ve hedef değişkeni (y) tanımla
X = data_cleaned.drop(['price'], axis=1)
y = data_cleaned['price']

Tahmin etmek istediğimiz değer price (fiyat) olduğu için data dan price sütunu çıkarıp X değişkenine atıyoruz.
Y değişkenine yani hedef değişkenimize price sütununun değerlerini atıyoruz.

In [ ]:
from sklearn.model_selection import train_test_split

# Eğitim ve test verisi olarak ayır
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=0)

Verilerimizi %70 eğitim ve %30 test olarak 2 ye ayırıyoruz. 

### Basit Model

In [ ]:
# Ortalama fiyat tahmin modeli
mean_model = np.full_like(y_test, y_test.mean(), dtype=np.float64)
# Medyan fiyat tahmin modeli
median_model = np.full_like(y_test, y_test.median(), dtype=np.float64)

median_mae, median_mse, median_rmse, median_mape, median_r2 = error_calculator(y_test, median_model)
print(f"Median Model mse: {median_mse}")
print(f"Median Model RMSE: {median_rmse}")
print(f"Median Model mape: {round(median_mape)} %")
print(f"Median Model r2: {median_r2}")
print("-" * 25)
mean_mae, mean_mse, mean_rmse, mean_mape, mean_r2 = error_calculator(y_test, mean_model)
print(f"Mean Model mse: {mean_mse}")
print(f"Mean Model RMSE: {mean_rmse}")
print(f"Mean Model mape: {round(mean_mape)} %")
print(f"Mean Model r2: {mean_r2}")

In [ ]:
pred_plot(y_test, median_model, "Median Model")
pred_plot(y_test, mean_model, "Mean Model")

### Linear Regression

In [ ]:
from sklearn.linear_model import LinearRegression

# Linear Regression modeli eğit
lr_model = LinearRegression()
lr_model.fit(X_train, y_train)

In [ ]:
y_lr_pred = lr_model.predict(X_test)

In [ ]:
mae, mse, rmse, mape, r2 = error_calculator(y_test, y_lr_pred)

print(f"Linear Regression mse: {mse}")
print(f"Linear Regression RMSE: {rmse}")
print(f"Linear Regression mape: {round(mape)} %")
print(f"Linear Regression r2: {r2}")

In [ ]:
pred_plot(y_test, y_lr_pred, "Linear Regression")

### Random Forest

In [ ]:
from sklearn.ensemble import RandomForestRegressor

# Random Forest Regressor modeli eğit
rf_model = RandomForestRegressor(n_estimators=100, random_state=42)
rf_model.fit(X_train, y_train)

In [ ]:
y_rf_pred = rf_model.predict(X_test)

In [ ]:
mae, mse, rmse, mape, r2= error_calculator(y_test, y_rf_pred)

print(f"Random Forest mse: {mse}")
print(f"Random Forest RMSE: {rmse}")
print(f"Random Forest mape: {round(mape)} %")
print(f"Random Forest r2: {r2}")

In [ ]:
pred_plot(y_test, y_rf_pred, "Random Forest")

### Lightgbm

In [ ]:
import lightgbm as lgb

lgb_model = lgb.LGBMRegressor()
lgb_model.fit(X_train, y_train)
y_lightgbm_pred = lgb_model.predict(X_test)

In [ ]:
mae, mse, rmse, mape, r2 = error_calculator(y_test, y_lightgbm_pred)

print(f"lightgbm mse: {mse}")
print(f"lightgbm RMSE: {rmse}")
print(f"lightgbm mape: {round(mape)} %")
print(f"lightgbm r2: {r2}")

In [ ]:
pred_plot(y_test, y_lightgbm_pred, "Lightgbm")

### Gradient Boosting Regressor

In [ ]:
%%time
from sklearn.model_selection import RandomizedSearchCV
from sklearn.ensemble import GradientBoostingRegressor

# Gradient Boosting Regressor
gbr_model = GradientBoostingRegressor()
param_distributions = {
    'n_estimators': [100, 200],
    'learning_rate': [0.01, 0.1],
    'max_depth': [3, 4],
    'min_samples_split': [2, 5],
    'min_samples_leaf': [1, 2]
}

random_search = RandomizedSearchCV(estimator=gbr_model, param_distributions=param_distributions, n_iter=10, cv=5, n_jobs=-1, verbose=2, random_state=42)
random_search.fit(X_train, y_train)

best_gbr = random_search.best_estimator_
gbr_y_pred = best_gbr.predict(X_test)


mae, mse, xgb_rmse, xgb_mape, _ = error_calculator(y_test, gbr_y_pred)

print(f"Gradient Boosting Regression RMSE: {xgb_rmse}")
print(f"Gradient Boosting Regression mape: {round(xgb_mape)} %")
print(f'En İyi Hiperparametreler: {random_search.best_params_}')
print(f'En İyi Skor: {random_search.best_score_}')

### XGBoost Regressor

In [ ]:
import xgboost as xgb
# XGBoost Regressor
xgb_model = xgb.XGBRegressor(objective='reg:squarederror')
xgb_model.fit(X_train, y_train)
xgb_y_pred = xgb_model.predict(X_test)

mae, mse, xgb_rmse, xgb_mape, _ = error_calculator(y_test, xgb_y_pred)

# Modelleri Karşılaştırma

In [ ]:
# Ortalama model hataları
_, _, rmse_mean, mape_mean, r2_mean = error_calculator(y_test, mean_model)

# Medyan model hataları
_, _, rmse_median, mape_median, r2_median = error_calculator(y_test, median_model)

# Doğrusal regresyon hataları
_, _, rmse_linear, mape_linear, r2_linear = error_calculator(y_test, y_lr_pred)

# Decision Tree hataları
_, _, rmse_tree, mape_tree, r2_tree = error_calculator(y_test, y_rf_pred)

# Lightgbm hataları
_, _, rmse_lightgbm, mape_lightgbm, r2_lightgbm = error_calculator(y_test, y_lightgbm_pred)

# Gradient Boosting Regressor hataları
_, _, rmse_gbr, mape_gbr, r2_gbr = error_calculator(y_test, gbr_y_pred)

# XGBRegressor hataları
_, _, rmse_xgb, mape_xgb, r2_xgb = error_calculator(y_test, xgb_y_pred)

In [ ]:
rmse_values = pd.DataFrame([{
    "rmse": rmse_mean,
    "r2": r2_mean,
    "model": "Mean Model"
}, {
    "rmse": rmse_median,
    "r2": r2_median,
    "model": "Median Model"
}, {
    "rmse": rmse_linear ,
    "r2": r2_linear,
    "model": "Linear Model"
}, {
    "rmse": rmse_tree,
    "r2": r2_tree,
    "model": "Decision Tree Model"
}, {
    "rmse": rmse_lightgbm,
    "r2": r2_lightgbm,
    "model": "Lightgbm Model"
}, {
    "rmse": rmse_gbr,
    "r2": r2_gbr,
    "model": "Gradient Boosting Regressor Model"
}, {
    "rmse": rmse_xgb,
    "r2": r2_xgb,
    "model": "XGBRegressor Model"
}])

rmse_values = rmse_values.sort_values(by="rmse").reset_index().drop(["index"], axis=1)

In [ ]:
# Grafik oluşturma
fig, ax1 = plt.subplots(figsize=(17, 8))

# RMSE için çubuk grafiği
ax1.set_xlabel('Models')
ax1.set_ylabel('RMSE', color='tab:blue')
bars = ax1.bar(rmse_values["model"], rmse_values["rmse"], color='tab:red', alpha=0.6, label='RMSE')
ax1.tick_params(axis='y', labelcolor='tab:red')

for i in range(len(bars)):
    bar = bars[i]
    yval = bar.get_height()
    ax1.text(bar.get_x() + bar.get_width()/2, yval, str(round(rmse_values["r2"][i] * 100, ndigits=2)) + " %", va='bottom', ha='center', color='green')


# Başlık ve gösterge ekleme
fig.suptitle('Model Performans Karşılaştırması')
fig.tight_layout()
plt.show()

Veri setimizi çoğalttığımız zaman XGBRegressor Model ve Gradient Boosting Regressor modellerinin büyük veri setlerinde daha iyi çalıştığını gözlemledik.